### PR TITLE
Enable dense x bsc mm/addmm

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85379
* __->__ #85308

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @bhosmer

### CODE REVIEW

File: test/test_sparse_csr.py [hunk 1]
```diff
@@ -1338,6 +1338,20 @@ def run_test_block_addmm_addmv(self, addmv_addmm, c, a, b, op_b=False, op_out=Fa
...
+            c_t = c.transpose(-2, -1)
+            actual_t = addmv_addmm(c_t, b_t, a_t, alpha=alpha, beta=beta)
+            out_t = torch.empty_like(out.transpose(-2, -1))
+            addmv_addmm(c_t, b_t, a_t, alpha=alpha, beta=beta, out=out_t)
+
+            expected_t = alpha * (b.cpu().resolve_conj().numpy().transpose() *
+                                  a_bsr.transpose()) + beta * c.cpu().numpy().transpose()
+            self.assertEqual(actual_t, out_t)
+            self.assertEqual(actual_t, expected_t)
+
```
Comment (nikitaved) [inline test/test_sparse_csr.py:1354]: Can't we just merge these tests with the ones above? We could provide a function that does the transposition when needed.

File: test/test_sparse_csr.py [hunk 2]
```diff
@@ -1338,6 +1338,20 @@ def run_test_block_addmm_addmv(self, addmv_addmm, c, a, b, op_b=False, op_out=Fa
...
+            # test dense x bsc by transposing the entire test case
+            a_t = a.transpose(-2, -1)
+            b_t = b.transpose(-2, -1)
+            c_t = c.transpose(-2, -1)
+            actual_t = addmv_addmm(c_t, b_t, a_t, alpha=alpha, beta=beta)
+            out_t = torch.empty_like(out.transpose(-2, -1))
+            addmv_addmm(c_t, b_t, a_t, alpha=alpha, beta=beta, out=out_t)
+
+            expected_t = alpha * (b.cpu().resolve_conj().numpy().transpose() *
+                                  a_bsr.transpose()) + beta * c.cpu().numpy().transpose()
```
Comment (pearu) [inline test/test_sparse_csr.py:1351]: Should this be:
```suggestion
            expected_t = alpha * (b.cpu().resolve_conj().numpy().transpose() @
                                  a_bsr.transpose()) + beta * c.cpu().numpy().transpose()
```
?

File: test/test_sparse_csr.py [hunk 3]
```diff
@@ -1338,6 +1338,20 @@ def run_test_block_addmm_addmv(self, addmv_addmm, c, a, b, op_b=False, op_out=Fa
         self.assertEqual(actual, out)
         self.assertEqual(actual, expected)
 
+        if addmv_addmm == torch.addmm:
+            # test dense x bsc by transposing the entire test case
+            a_t = a.transpose(-2, -1)
+            b_t = b.transpose(-2, -1)
+            c_t = c.transpose(-2, -1)
+            actual_t = addmv_addmm(c_t, b_t, a_t, alpha=alpha, beta=beta)
+            out_t = torch.empty_like(out.transpose(-2, -1))
```
Comment (pearu) [inline test/test_sparse_csr.py:1347]: Nit:
```suggestion
            out_t = torch.empty_like(out).transpose(-2, -1)
```

**Review by pearu — COMMENTED** [review]
I am slightly confused by the test case that uses element-wise `mul` to test the result of `addmm`. Either the test is not enabled or the inputs to the tests are trivial (say, all zeros).

File: test/test_sparse_csr.py (hunk 2 above)
Comment (amjames) [reply test/test_sparse_csr.py:1351]: I was confused by this too, it appears that for `scipy.sparse.sp_matrix` subclasses (which `b_bsr` is `__mul__` is matrix multiplication. 

I will see about changing it.

File: test/test_sparse_csr.py [hunk 4]
```diff
@@ -1345,25 +1359,59 @@ def run_test_block_addmm_addmv(self, addmv_addmm, c, a, b, op_b=False, op_out=Fa
...
+                if out is not None:
+                    # the ref takes no out kwarg
+                    assert isinstance(out, torch.Tensor)
+                    # tranpose inplace to propogate out to checking context
+                    out.transpose_(-2, -1)
+                    return f(tt(c), tt(b), tt(a), alpha=alpha, beta=beta, out=out)
+                else:
+                    return f(tt(c), tt(b), tt(a), alpha=alpha, beta=beta)
+
+            return wrapper
```
Comment (amjames) [inline test/test_sparse_csr.py:1385]: @nikitaved This solution using a wrapper which generates the transposed + permuted op is the closest to your suggestion I could come up with.

**Review by pearu — APPROVED** [review]
LGTM!

File: test/test_sparse_csr.py [hunk 5]
```diff
@@ -1494,25 +1508,59 @@ def run_test_block_addmm_addmv(self, addmv_addmm, c, a, b, op_b=False, op_out=Fa
...
                 a = a.to_sparse_bsr((block_size, block_size))
             else:
                 a = self.genSparseCSRTensor((m, k), nnz, dtype=dtype, device=device, index_dtype=index_dtype)
                 a_data = make_tensor((nnz, block_size, block_size), dtype=dtype, device=device)
-                a_data = a_data.mT if noncontiguous else a_data   # Test column-major blocks
-                a = torch._sparse_bsr_tensor_unsafe(a.crow_indices(), a.col_indices(),
-                                                    a_data, (m * block_size, k * block_size))
+                a_data = a_data.mT if noncontiguous else a_data  # Test column-major blocks
+                a = torch._sparse_bsr_tensor_unsafe(a.crow_indices(), a.col_indices(), a_data,
+                                                    (m * block_size, k * block_size))
```
Comment (nikitaved) [inline test/test_sparse_csr.py:1550]: Yeah, @amjames , we will have a conflict here with https://github.com/pytorch/pytorch/pull/85551. We can created tensors unconditionally and just transpose `a_data` if needed.

Comment (amjames) [reply test/test_sparse_csr.py:1550]: This should be landed today if you want to just wait and rebase?

**Pull request closed without merging**